File: almost-equal-powers.pl
#!/usr/bin/perl

# find powers of different bases which are approximately equal
# e.g. 4**64=340282366920938463463374607431768211456 ~~ 11**37=340039485861577398992406882305761986971 (0.0714%)

use strict;
use warnings;

use POSIX qw(floor ceil);
use bignum;

my @bases = map { Math::BigInt->new($_) } (2, 3, 5, 7, 11);
my @powers = map { Math::BigInt->new($_) } (1 .. 100);
my $low_tolerance = 1.01;
my $high_tolerance = 1 / $low_tolerance;

for my $base1 (@bases) {
    for my $power1 (@powers) {
        my $big1 = pow($base1, $power1);
        for my $base2 (@bases) {
            next unless $base2 > $base1 && # don't duplicate work
                        Math::BigInt::bgcd($base1, $base2) == 1; # only do for relative primes
            my $power2 = $power1 * log($base1) / log($base2);
            my $power2_low = floor $power2;
            my $power2_high = ceil $power2;
            my $big2_low = pow($base2, $power2_low);
            my $big2_high = pow($base2, $power2_high);

            my $ratio_low = $big1 / $big2_low;
            if ($ratio_low < $low_tolerance) {
                printf "%d**%d=%s ~~ %d**%d=%s (%.4f%%)\n",
                    $base1, $power1, $big1->bstr(), $base2, $power2_low, $big2_low->bstr(), 100 * ($ratio_low-1);
            }

            my $ratio_high = $big1 / $big2_high;
            if ($ratio_high > $high_tolerance) {
                printf "%d**%d=%s ~~ %d**%d=%s (%.4f%%)\n",
                    $base1, $power1, $big1->bstr(), $base2, $power2_high, $big2_high->bstr(), 100 * ($ratio_high-1);
            }
        }
    }
}

my %cache;
sub pow {
    my ($base, $power) = @_;
    return $cache{"$base**$power"} //= $base ** $power;
}

File: almost-equal-powers.go
package main

import (
	"flag"
	"fmt"
	"math"
	"math/big"
)

func main() {
	var (
		minBase, maxBase, minPower, maxPower int
		tolerance                            float64
		primeBases, verbose, debug           bool
	)
	flag.IntVar(&minBase, "min-base", 2, "minimum base")
	flag.IntVar(&maxBase, "max-base", 100, "maximum base")
	flag.IntVar(&minPower, "min-power", 1, "minimum power")
	flag.IntVar(&maxPower, "max-power", 200, "maximum power")
	flag.Float64Var(&tolerance, "tolerance", 0.00001, "tolerance for search")
	flag.BoolVar(&primeBases, "prime-bases", false, "consider only prime bases")
	flag.BoolVar(&verbose, "verbose", false, "show all calculations")
	flag.BoolVar(&debug, "debug", false, "show debugging info")
	flag.Parse()

	lowTolerance, highTolerance := big.NewRat(0, 1), big.NewRat(0, 1)
	lowTolerance.SetFloat64(1 + tolerance)
	highTolerance.Inv(lowTolerance)

	bases := make([]*big.Int, 0)
	for i := minBase; i <= maxBase; i++ {
		base := big.NewInt(int64(i))
		if !primeBases || base.ProbablyPrime(16) {
			bases = append(bases, base)
		}
	}
	if debug {
		fmt.Printf("bases=%v\n", bases)
	}
	powers := make([]*big.Int, 0, maxPower-minPower+1)
	for i := minPower; i <= maxPower; i++ {
		powers = append(powers, big.NewInt(int64(i)))
	}

	type tuple struct {
		base, power *big.Int
	}
	hits, misses := 0, 0
	cache := make(map[tuple]*big.Int)
	pow := func(base, power *big.Int) *big.Int {
		t := tuple{base, power}
		if value, exists := cache[t]; exists {
			hits++
			return value
		}
		misses++
		value := big.NewInt(0).Exp(base, power, nil)
		cache[t] = value
		return value
	}

	oneInt := big.NewInt(1)

	for _, base1 := range bases {
		log1 := math.Log(float64(base1.Int64()))
		for _, base2 := range bases {
			// don't duplicate work
			if base1.Cmp(base2) >= 0 {
				continue
			}

			// only consider relative primes
			var gcd big.Int
			gcd.GCD(nil, nil, base1, base2)
			if gcd.Cmp(oneInt) != 0 {
				continue
			}

			for _, power1 := range powers {
				big1 := pow(base1, power1)
				log2 := math.Log(float64(base2.Int64()))

				power2 := float64(power1.Int64()) * log1 / log2
				power2Low := big.NewInt(int64(math.Floor(power2)))
				power2High := big.NewInt(int64(math.Ceil(power2)))

				big2Low := pow(base2, power2Low)
				big2High := pow(base2, power2High)

				if verbose {
					fmt.Printf("%v**%v (%v) < %v**%v (%v) < %v**%v (%v)\n",
						base2, power2Low, big2Low,
						base1, power1, big1,
						base2, power2High, big2High,
					)
				}

				ratioLow := big.NewRat(0, 1).SetFrac(big1, big2Low)
				ratioHigh := big.NewRat(0, 1).SetFrac(big1, big2High)

				if ratioLow.Cmp(lowTolerance) < 0 {
					rat, _ := ratioLow.Float64()
					fmt.Printf("%v**%v=%v ~~ %v**%v=%v (%.6f%%)\n",
						base1, power1, big1,
						base2, power2Low, big2Low,
						100*(rat-1),
					)
				}

				if ratioHigh.Cmp(highTolerance) > 0 {
					rat, _ := ratioHigh.Float64()
					fmt.Printf("%v**%v=%v ~~ %v**%v=%v (%.6f%%)\n",
						base1, power1, big1,
						base2, power2High, big2High,
						100*(rat-1),
					)
				}
			}
		}
	}

	if debug {
		fmt.Printf("Cache: %d/%d=%.1f%%\n", hits, misses, 100*float64(hits)/(float64(hits+misses)))
	}
}
